src/server/game/Entities/Player/Player.cpp

```diff
@@ -1540,27 +1541,40 @@ bool Player::BuildEnumData(PreparedQueryResult result, WorldPacket* data)
     *data << uint32(petLevel);
     *data << uint32(petFamily);
 
-    Tokenizer equipment(fields[22].GetString(), ' ');
+    std::vector<std::string_view> equipment = Trinity::Tokenize(fields[22].GetString(), ' ', false);
     for (uint8 slot = 0; slot < INVENTORY_SLOT_BAG_END; ++slot)
     {
         uint32 visualBase = slot * 2;
-        uint32 itemId = GetUInt32ValueFromArray(equipment, visualBase);
-        ItemTemplate const* proto = sObjectMgr->GetItemTemplate(itemId);
-        if (!proto)
+        if (!((visualBase+1) < equipment.size()))
         {
             *data << uint32(0);
             *data << uint8(0);
             *data << uint32(0);
             continue;
         }
 
+        ItemTemplate const* proto = nullptr;
+        if (Optional<uint32> itemId = Trinity::StringTo<uint32>(equipment[visualBase]))
+            proto = sObjectMgr->GetItemTemplate(*itemId);
+
+        if (!proto)
+        {
+            TC_LOG_ERROR("entities.player.loading", "Player %u has invalid equipment '%s' in `equipmentcache`. Load aborted.", guid, std::string(equipment[visualBase]).c_str());
+            return false;
```

Comment:
Do not skip building character enum for a silly reason like this, just send empty slot like it used to

Reply:
This one I am a bit more conflicted on, but generally I feel that if character data is invalid, I'd rather fail safely (refuse to load) rather than fail deadly (potentially lose data somewhere down the line).

Reply:
Except deleting invalid items is the intended behavior here